### PR TITLE
API - Requester - User owns a task

### PR DESCRIPTION
**Pull request by Carltesio**
# [PT Story](https://www.pivotaltracker.com/story/show/172275248)
```
As a person in need of help
in order to associate a help request to me
I would like the help request to be associated to my profile
```
## Changes proposed in this pull request:
*  Associates a profile to a specific task
* Refactors the tests
* Adds max and min amount of items requestable
## What I have learned working on this feature:
* How to refactor the test using better the factory bots
* Oliver knows.

### CODE REVIEW

File: db/migrate/20200415145622_add_task_to_user.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class AddTaskToUser < ActiveRecord::Migration[6.0]
+  def change
+    change_table :tasks do |t|
+      t.belongs_to :user, null: false, foreign_key: true
```
Comment (robin-lillqvist) [inline db/migrate/20200415145622_add_task_to_user.rb:4]: What exactly is this doing? 
Is it adding a task to user table?
or user id to task table??

File: app/controllers/api/v1/tasks_controller.rb [hunk 2]
```diff
@@ -1,32 +1,55 @@
...
+      render json: create_json_response(task)    
     end
   end
 
   private
 
+  def render_error_message(task)
+    case 
+    when task.task_items.count >= 40
+      message = 'You have to many products selected.'
```
Comment (pierre-1) [inline app/controllers/api/v1/tasks_controller.rb]: There are a few typos and we should address

File: app/controllers/api/v1/tasks_controller.rb [hunk 3]
```diff
@@ -1,32 +1,55 @@
...
+      message = 'You have to pick atleast 5 products.'
+    else 
+      message = 'Internal problem. Contact support. No activity specified'
+    end
+
+    render json: {error_message: message}, status: 400
+  end
+
+  def render_active_record_error(error)
+    render json: {error_message: "Internal problem. Contact support. #{error.message}"}, status: 400
```
Comment (pierre-1) [inline app/controllers/api/v1/tasks_controller.rb]: In what case are the users seeing these error messages? If they can potentially encounter them, the messages should probably be phrased in a more user-friendly way.

File: db/migrate/20200415145622_add_task_to_user.rb (hunk 1 above)
Comment (JaimeCrz) [reply db/migrate/20200415145622_add_task_to_user.rb:4]: Task will belong to user, 

User_id in Task table.

File: app/controllers/api/v1/tasks_controller.rb (hunk 2 above)
Comment (JaimeCrz) [reply app/controllers/api/v1/tasks_controller.rb]: Will fix this in a couple of hours!

File: app/controllers/api/v1/tasks_controller.rb (hunk 3 above)
Comment (JaimeCrz) [reply app/controllers/api/v1/tasks_controller.rb]: We will double check on that, it's due active record messages. 

This error message it is practically impossible to happen but it's when you send in an invalid Task number with your request, Aka trying to update an order with postman that doesn't belong to you.

File: app/controllers/api/v1/tasks_controller.rb [hunk 4]
```diff
@@ -25,6 +32,30 @@ def update
 
   private
 
+  def restrict_user_to_have_one_active_task
+    if current_user.tasks.any?{|task| task.status == 'pending' || 'confirmed'}
+      render json: { error: "You already have an active task pending" }, status: 403
```
Comment (robin-lillqvist) [inline app/controllers/api/v1/tasks_controller.rb]: This error message could be more user friendly

File: app/models/task.rb [hunk 5]
```diff
@@ -1,3 +1,5 @@
 class Task < ApplicationRecord
   has_many :task_items
+  belongs_to :user
+  enum status: [:pending, :confirmed, :claimed, :finalized]
```
Comment (pierre-1) [inline app/models/task.rb:4]: Sounds logical, I just would like to double check the meaning of each status

File: app/controllers/api/v1/tasks_controller.rb [hunk 6]
```diff
@@ -25,6 +32,30 @@ def update
...
+      return
+    end
+  end
+
+  def render_error_message(task)
+    case
+    when task.task_items.count >= 40
+      message = "You have to many products selected."
```
Comment (blake-futchi) [inline app/controllers/api/v1/tasks_controller.rb]: you have selected too many products

File: app/controllers/api/v1/tasks_controller.rb (hunk 4 above)
Comment (robin-lillqvist) [reply app/controllers/api/v1/tasks_controller.rb]: You can only have one active task active at one time

File: app/controllers/api/v1/tasks_controller.rb [hunk 7]
```diff
@@ -25,6 +32,30 @@ def update
...
+    when task.task_items.count < 5
+      message = "You have to pick at least 5 products."
+    else
+      message = "Internal problem. Contact support. No activity specified"
```
Comment (blake-futchi) [inline app/controllers/api/v1/tasks_controller.rb]: We are experiencing internal errors. Please refresh the page and contact support.

File: spec/requests/api/v1/user_can_see_list_of_tasks_spec.rb [hunk 8]
```diff
@@ -1,15 +1,14 @@
...
+  let!(:confirmed_task_items) { 5.times { create(:task_item, task: confirmed_task) } }
 
-  before do
-    post "/api/v1/tasks", params: { product_id: product.id }
-    @task_id = (response_json)["task"]["id"]
-    @task = Task.find(@task_id)
-    put "/api/v1/tasks/#{@task.id}", params: { activity: "confirmed" }
-  end
+  let(:another_confirmed_task) { create(:task, status: 'confirmed') }
+  let!(:another_confirme_task_items) { 5.times { create(:task_item, task: another_confirmed_task) } }
```
Comment (blake-futchi) [inline spec/requests/api/v1/user_can_see_list_of_tasks_spec.rb]: typo should be confirmed task?

Comment (blake-futchi) [reply spec/requests/api/v1/user_can_see_list_of_tasks_spec.rb]: how is it passing with this typo?

**Review by blake-futchi — COMMENTED** [review]
check comments

File: spec/requests/api/v1/user_creates_task_spec.rb [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      it "response with the right total amount" do
+        expect(response_json["task"]["total"]).to eq "110.0"
+      end
+    end
+  end
+
+  describe "Unsuccessfull" do
+    describe "when user has a pending task" do
+      let(:pending_task) { create(:task, user: user, status: 'pending') }
+      let!(:task_items) { 5.times { create(:task_item, task: pending_task) } }
```
Comment (robin-lillqvist) [inline spec/requests/api/v1/user_creates_task_spec.rb]: This is problematic if the user cannot re-use the pending task that is not confirmed

File: app/controllers/api/v1/tasks_controller.rb (hunk 2 above)
Comment (JaimeCrz) [reply app/controllers/api/v1/tasks_controller.rb]: done

File: app/controllers/api/v1/tasks_controller.rb (hunk 7 above)
Comment (JaimeCrz) [reply app/controllers/api/v1/tasks_controller.rb]: done

File: app/controllers/api/v1/tasks_controller.rb (hunk 4 above)
Comment (JaimeCrz) [reply app/controllers/api/v1/tasks_controller.rb]: done

**Review by robin-lillqvist — APPROVED** [review]
LGTM

At first github was fucking with me and didn't show the changes. But now I see them

**Pull request merged**